README.md
Node.js Long-Emitter
===

A buffered event emitter designed for use in simple long-polling event streaming from Node.js servers to web site front-ends, and managing delayed responses & timeouts.

[![NPM version](https://img.shields.io/npm/v/long-emitter.svg)](http://npmjs.org/long-emitter)
[![Licensed MIT](https://img.shields.io/npm/l/long-emitter.svg)](https://github.com/ChiperSoft/node-long-emitter/blob/master/LICENSE.txt)
[![Nodejs 0.10+](https://img.shields.io/badge/node.js-%3E=_0.10-brightgreen.svg)](http://nodejs.org)
[![Downloads](http://img.shields.io/npm/dm/long-emitter.svg)](http://npmjs.org/long-emitter)
[![Build Status](https://img.shields.io/travis/ChiperSoft/node-long-emitter.svg)](https://travis-ci.org/ChiperSoft/node-long-emitter)
[![Coverage Status](https://img.shields.io/coveralls/ChiperSoft/node-long-emitter.svg)](https://coveralls.io/r/ChiperSoft/node-long-emitter)
[![Gittip](http://img.shields.io/gittip/chipersoft.svg)](https://www.gittip.com/chipersoft/)

#Installation

NPM: `npm install long-emitter`

##Example

See the [example application](https://github.com/ChiperSoft/node-long-emitter/tree/master/example) for a basic use case.

##Basic Usage

```js
var longEmitter = require('long-emitter');
var manager = longEmitter();
var emitter = manager.create();
emitter.emit('a');
emitter.emit('b', 2);
emitter.drain(function (events) {
	// events -> [ ['a'], ['b', 2], ['c', 'hello'] ]
});
emitter.emit('c', 'hello');
emitter.release();
```

##Usage

Documentation to be written.

History.md
0.2.2 / 2014-11-20
==================

  * Added History.md and README.md
  * Changed "server" to "manager" to make it clearer what it does.



tests/bucket.js
var Bucket = require('../lib/bucket.js');
var EventEmitter = require('events').EventEmitter;

exports['emits to queue'] = function (test) {
	var b = new Bucket();
	b.emit('testevent', 'arg1', 'argtwo');

	test.deepEqual(b._queue, [['testevent', 'arg1', 'argtwo']]);

	test.done();
};

exports['multiple emits to queue'] = function (test) {
	var b = new Bucket();
	b.emit('testevent', 'arg1', 'argtwo');
	b.emit('onemore', 2, {foo:'bar'});

	test.deepEqual(b._queue, [['testevent', 'arg1', 'argtwo'], ['onemore', 2, {foo:'bar'}]]);

	test.done();
};

exports['drains into callback'] = function (test) {
	var b = new Bucket();
	b.emit('testevent', 'arg1', 'argtwo');
	b.emit('onemore', 2, {foo:'bar'});

	b.drain(function (events) {
		test.deepEqual(events, [['testevent', 'arg1', 'argtwo'], ['onemore', 2, {foo:'bar'}]]);

		test.done();
	});
	
};

exports['drains into promise'] = function (test) {
	var b = new Bucket();
	b.emit('testevent', 'arg1', 'argtwo');
	b.emit('onemore', 2, {foo:'bar'});

	b.drain().then(function (events) {
		test.deepEqual(events, [['testevent', 'arg1', 'argtwo'], ['onemore', 2, {foo:'bar'}]]);

		test.done();
	});
	
};

exports['drain emits _drained'] = function (test) {
	var b = new Bucket();
	b.emit('one');
	b.emit('two');
	b.emit('three');

	var emitted = 0;

	b.once('_drained', function () {
		emitted++;
		test.strictEqual(emitted, 2);

		test.done();
	});

	b.drain(function () {
		emitted++;
		test.strictEqual(emitted, 1);

	});
};


exports['drains async'] = function (test) {
	var b = new Bucket();
	b.drain(function (events) {
		test.deepEqual(events, [['testevent', 'arg1', 'argtwo'], ['onemore', 2, {foo:'bar'}]]);

		test.done();
	});

	b.emit('testevent', 'arg1', 'argtwo');

	b.emit('onemore', 2, {foo:'bar'});
	
};

exports['drains async, with pause'] = function (test) {
	test.expect(2);

	var b = new Bucket({pause: 10, maxPause: 50});
	b.drain(function (events) {
		test.deepEqual(events, [['testevent', 'arg1', 'argtwo'], ['onemore', 2, {foo:'bar'}]]);

	});

	b.emit('testevent', 'arg1', 'argtwo');

	b.emit('onemore', 2, {foo:'bar'});

	setTimeout(function () {
		b.emit('three');

		b.drain(function (events) {
			test.deepEqual(events, [['three']]);
			test.done();
		});
	}, 100);
	
};

exports['drains async, with max wait'] = function (test) {
	test.expect(2);

	var b = new Bucket({pause: 100, maxPause: 500});

	b.drain(function (events) {
		test.deepEqual(events, [
			['iteration', 10],
			['iteration', 9],
			['iteration', 8],
			['iteration', 7],
			['iteration', 6],
			['iteration', 5]
		]);

	});

	var count = 10;
	function loop () {
		b.emit('iteration', count--);

		if (count > 0) setTimeout(loop, 90);
	}
	loop();

	setTimeout(function () {
		b.drain(function (events) {
			test.deepEqual(events, [
				['iteration', 4],
				['iteration', 3],
				['iteration', 2],
				['iteration', 1]
			]);
			test.done();
		});
	}, 1000);

	
};

exports['tapped into emitter'] = function (test) {
	test.expect(4);

	var b = new Bucket({pause: 10, maxPause: 50});

	b.emit('testevent', 'arg1', 'argtwo');

	b.emit('onemore', 2, {foo:'bar'});

	var emitter = new EventEmitter();
	emitter.on('testevent', function (arg1, arg2) {
		test.strictEqual(arg1, 'arg1');
		test.strictEqual(arg2, 'argtwo');
	});
	emitter.on('onemore', function (arg1, arg2) {
		test.strictEqual(arg1, 2);
		test.deepEqual(arg2, {foo: 'bar'});
	});
	emitter.on('three', function () {
		test.ok(false, 'should not have received third event');
	});

	emitter.on('_empty', test.done);
	
	b.tap(emitter);

	b.emit('three');
};

exports['tap error'] = function (test) {
	var b = new Bucket();
	b.emit('testevent', 'arg1', 'argtwo');

	test.throws(function () {
		b.tap(false);
	});

	test.done();
};

exports['drain timeout'] = function (test) {
	test.expect(2);
	var b = new Bucket({maxIdle: 500});
	var waited = false;

	setTimeout(function () {waited = true;}, 490);

	b.drain(function (events) {
		test.deepEqual(events, []);
		test.ok(waited, 'Did not wait');
		test.done();
	});
};

exports['drain abort'] = function (test) {
	test.expect(2);
	var b = new Bucket();

	var hook = b.drain(function (events) {
		test.ok(false, 'Still drained');
	});

	test.strictEqual(typeof hook, 'object');

	setTimeout(function () {
		hook.abort();

		b.emit('one');

		test.ok(true);
		test.done();
	}, 50);
};

exports['emit errors w/ listener'] = function (test) {
	test.expect(2);
	var b = new Bucket({maxIdle: 500});

	var e = new Error('Test error');

	b.on('error', function (err) {
		test.strictEqual(err, e);
	});

	b.drain(function (events) {
		test.deepEqual(events, [['one']]);
		test.done();
	});

	b.emit('error', e);
	b.emit('one');
};

exports['emit errors w/o listener'] = function (test) {
	test.expect(1);
	var b = new Bucket({maxIdle: 500});

	var e = new Error('Test error');

	b.drain(function (events) {
		test.deepEqual(events, [['error', e], ['one']]);
		test.done();
	});

	b.emit('error', e);
	b.emit('one');
};


tests/server.js
var longEmitter = require('../lib/server.js');
var EventEmitter = require('events').EventEmitter;

exports['initializes correctly'] = function (test) {
	var server = longEmitter();

	test.deepEqual(Object.keys(server), ['create','get','close','finish','param','query', 'clear']);

	test.done();
	server.clear(); //cleanup so our tests don't hang
};

exports['create bucket'] = function (test) {
	var server = longEmitter();

	var a = server.create();
	var id = a.id;

	test.ok(a);
	test.ok(id);

	var b = server.get(id);

	test.strictEqual(a, b);

	test.done();
	server.clear(); //cleanup so our tests don't hang
};

exports['close bucket'] = function (test) {
	var server = longEmitter();

	var a = server.create();
	var id = a.id;

	test.ok(a);
	test.ok(id);

	server.close(id);

	test.ok(!server.get(id));

	test.done();
	server.clear(); //cleanup so our tests don't hang
};

exports['bucket cleanup'] = function (test) {
	var server = longEmitter({
		expiration: 90,
		cleanupPeriod: 50
	});

	var a = server.create();
	var id = a.id;

	test.ok(a);
	test.ok(id);

	setTimeout(function () {
		var b = server.get(id);
		test.ok(!b, 'Bucket did not expire');
		test.done();
		server.clear(); //cleanup so our tests don't hang
	}, 125);
};

exports['bucket finished'] = function (test) {
	var server = longEmitter();

	var a = server.create();
	var id = a.id;

	a.emit('one');

	server.finish(id);
	
	a.emit('two');

	test.ok(server.get(id), 'I lost mah bucket.');

	a.drain();

	a.once('_drained', function () {
		test.ok(!server.get(id), 'Bucket still exists');

		test.done();

		server.clear(); //cleanup so our tests don't hang
	});

};

exports['bucket released'] = function (test) {
	var server = longEmitter();

	var a = server.create();
	var id = a.id;

	a.emit('one');

	a.release();
	
	a.emit('two');

	test.ok(server.get(id), 'I lost mah bucket.');

	a.drain();

	a.once('_drained', function () {
		test.ok(!server.get(id), 'Bucket still exists');

		test.done();

		server.clear(); //cleanup so our tests don't hang
	});
};

exports['param middleware'] = function (test) {
	var server = longEmitter();

	var a = server.create();

	a.emit('one');
	a.emit('two');
	a.emit('three');

	var res = {locals: {}};
	var req = new EventEmitter();

	server.param(req, res, function (err) {
		test.ok(!err);
		test.deepEqual(res, {locals: {events: [['one'],['two'],['three']]}});
		test.done();

		server.clear();
	}, a.id);
};

exports['param disconnect'] = function (test) {
	test.expect(1);
	var server = longEmitter();

	var a = server.create();

	var res = {locals: {}};
	var req = new EventEmitter();

	server.param(req, res, function (err) {
		test.ok(false, 'Request should not have continued');
	}, a.id);

	setTimeout(function () {
		test.ok(req.emit('close'));

		server.clear();
		test.done();
	}, 50);
};

exports['query middleware'] = function (test) {
	var server = longEmitter({
		query: 'request'
	});

	var a = server.create();

	a.emit('one');
	a.emit('two');
	a.emit('three');

	var req = new EventEmitter();
	req.query = {request: a.id};

	var res = {locals: {}};

	server.query(req, res, function (err) {
		test.ok(!err);
		test.deepEqual(res, {locals: {events: [['one'],['two'],['three']]}});
		test.done();

		server.clear();
	});
};


exports['query route'] = function (test) {
	var server = longEmitter({
		query: function (req) {
			return req.query.request;
		},
		querySend: true
	});

	var a = server.create();

	a.emit('one');
	a.emit('two');
	a.emit('three');

	var req = new EventEmitter();
	req.query = {request: a.id};

	var res = {locals: {}, json: function (body) {
		test.deepEqual(body, [['one'],['two'],['three']]);
		test.done();

		server.clear();
	}};

	server.query(req, res, function (err) {
		test.ok(false, 'Should not have called next()');
	});
};

exports['query disconnect'] = function (test) {
	test.expect(1);
	var server = longEmitter({
		query: function (req) {
			return req.query.request;
		},
		querySend: true
	});

	var a = server.create();

	var req = new EventEmitter();
	req.query = {request: a.id};

	var res = {locals: {}, json: function (body) {
		test.ok(false, 'Should not have called res.json()');
	}};

	server.query(req, res, function (err) {
		test.ok(false, 'Request should not have continued');
	}, a.id);

	setTimeout(function () {
		test.ok(req.emit('close'));

		server.clear();
		test.done();
	}, 50);
};

exports['multiple emitters'] = function (test) {
	test.expect(2);
	var server = longEmitter();
	var a = server.create();
	var b = server.create();

	setTimeout(function () {
		a.emit('one');
	}, 10);

	setTimeout(function () {
		b.emit('two');
	}, 20);

	setTimeout(function () {
		b.emit('three');
		server.finish(b.id);
	}, 30);

	setTimeout(function () {
		var b2 = server.get(b.id);
		b2.drain(function (results) {
			test.deepEqual(results, [['two'], ['three']]);
		});
	}, 40);

	setTimeout(function () {
		a.emit('four');
		server.finish(a);
	}, 50);

	setTimeout(function () {
		var a2 = server.get(a.id);
		a2.drain(function (results) {
			test.deepEqual(results, [['one'], ['four']]);
			test.done();

			server.clear();
		});
	}, 60);
};


tests/manager.js
var longEmitter = require('../lib/manager.js');
var EventEmitter = require('events').EventEmitter;

exports['initializes correctly'] = function (test) {
	var manager = longEmitter();

	test.deepEqual(Object.keys(manager), ['create','get','close','finish','param','query', 'clear']);

	test.done();
	manager.clear(); //cleanup so our tests don't hang
};

exports['create bucket'] = function (test) {
	var manager = longEmitter();

	var a = manager.create();
	var id = a.id;

	test.ok(a);
	test.ok(id);

	var b = manager.get(id);

	test.strictEqual(a, b);

	test.done();
	manager.clear(); //cleanup so our tests don't hang
};

exports['close bucket'] = function (test) {
	var manager = longEmitter();

	var a = manager.create();
	var id = a.id;

	test.ok(a);
	test.ok(id);

	manager.close(id);

	test.ok(!manager.get(id));

	test.done();
	manager.clear(); //cleanup so our tests don't hang
};

exports['bucket cleanup'] = function (test) {
	var manager = longEmitter({
		expiration: 90,
		cleanupPeriod: 50
	});

	var a = manager.create();
	var id = a.id;

	test.ok(a);
	test.ok(id);

	setTimeout(function () {
		var b = manager.get(id);
		test.ok(!b, 'Bucket did not expire');
		test.done();
		manager.clear(); //cleanup so our tests don't hang
	}, 125);
};

exports['bucket finished'] = function (test) {
	var manager = longEmitter();

	var a = manager.create();
	var id = a.id;

	a.emit('one');

	manager.finish(id);
	
	a.emit('two');

	test.ok(manager.get(id), 'I lost mah bucket.');

	a.drain();

	a.once('_drained', function () {
		test.ok(!manager.get(id), 'Bucket still exists');

		test.done();

		manager.clear(); //cleanup so our tests don't hang
	});

};

exports['bucket released'] = function (test) {
	var manager = longEmitter();

	var a = manager.create();
	var id = a.id;

	a.emit('one');

	a.release();
	
	a.emit('two');

	test.ok(manager.get(id), 'I lost mah bucket.');

	a.drain();

	a.once('_drained', function () {
		test.ok(!manager.get(id), 'Bucket still exists');

		test.done();

		manager.clear(); //cleanup so our tests don't hang
	});
};

exports['param middleware'] = function (test) {
	var manager = longEmitter();

	var a = manager.create();

	a.emit('one');
	a.emit('two');
	a.emit('three');

	var res = {locals: {}};
	var req = new EventEmitter();

	manager.param(req, res, function (err) {
		test.ok(!err);
		test.deepEqual(res, {locals: {events: [['one'],['two'],['three']]}});
		test.done();

		manager.clear();
	}, a.id);
};

exports['param disconnect'] = function (test) {
	test.expect(1);
	var manager = longEmitter();

	var a = manager.create();

	var res = {locals: {}};
	var req = new EventEmitter();

	manager.param(req, res, function (err) {
		test.ok(false, 'Request should not have continued');
	}, a.id);

	setTimeout(function () {
		test.ok(req.emit('close'));

		manager.clear();
		test.done();
	}, 50);
};

exports['query middleware'] = function (test) {
	var manager = longEmitter({
		query: 'request'
	});

	var a = manager.create();

	a.emit('one');
	a.emit('two');
	a.emit('three');

	var req = new EventEmitter();
	req.query = {request: a.id};

	var res = {locals: {}};

	manager.query(req, res, function (err) {
		test.ok(!err);
		test.deepEqual(res, {locals: {events: [['one'],['two'],['three']]}});
		test.done();

		manager.clear();
	});
};


exports['query route'] = function (test) {
	var manager = longEmitter({
		query: function (req) {
			return req.query.request;
		},
		querySend: true
	});

	var a = manager.create();

	a.emit('one');
	a.emit('two');
	a.emit('three');

	var req = new EventEmitter();
	req.query = {request: a.id};

	var res = {locals: {}, json: function (body) {
		test.deepEqual(body, [['one'],['two'],['three']]);
		test.done();

		manager.clear();
	}};

	manager.query(req, res, function (err) {
		test.ok(false, 'Should not have called next()');
	});
};

exports['query disconnect'] = function (test) {
	test.expect(1);
	var manager = longEmitter({
		query: function (req) {
			return req.query.request;
		},
		querySend: true
	});

	var a = manager.create();

	var req = new EventEmitter();
	req.query = {request: a.id};

	var res = {locals: {}, json: function (body) {
		test.ok(false, 'Should not have called res.json()');
	}};

	manager.query(req, res, function (err) {
		test.ok(false, 'Request should not have continued');
	}, a.id);

	setTimeout(function () {
		test.ok(req.emit('close'));

		manager.clear();
		test.done();
	}, 50);
};

exports['multiple emitters'] = function (test) {
	test.expect(2);
	var manager = longEmitter();
	var a = manager.create();
	var b = manager.create();

	setTimeout(function () {
		a.emit('one');
	}, 10);

	setTimeout(function () {
		b.emit('two');
	}, 20);

	setTimeout(function () {
		b.emit('three');
		manager.finish(b.id);
	}, 30);

	setTimeout(function () {
		var b2 = manager.get(b.id);
		b2.drain(function (results) {
			test.deepEqual(results, [['two'], ['three']]);
		});
	}, 40);

	setTimeout(function () {
		a.emit('four');
		manager.finish(a);
	}, 50);

	setTimeout(function () {
		var a2 = manager.get(a.id);
		a2.drain(function (results) {
			test.deepEqual(results, [['one'], ['four']]);
			test.done();

			manager.clear();
		});
	}, 60);
};


lib/bucket.js
var debounce = require('lodash.debounce');
var assign = require('lodash.assign');
var proxmis = require('proxmis');
var EventEmitter = require('events').EventEmitter;
var debug  = require('debug')('long-emitter:bucket');


function Bucket (options) {
	EventEmitter.call(this);

	options = assign({
		pause: 20,
		maxPause: 500,
		maxIdle: 1000 * 30
	}, options || {});

	this._queue = [];
	this._drains = [];
	this._options = options;

	this.send = debounce(this.send, options.pause, {maxWait: options.maxPause});
	debug('created');
}

Bucket.prototype = Object.create(EventEmitter.prototype);

Bucket.prototype._emit = EventEmitter.prototype.emit;

Bucket.prototype.emit = function emit (name) {
	if (this._events[name] && this._emit.apply(this, arguments)) {
		debug('emitted to local', name);
		return true;
	}

	debug('emitted to queue', name);

	var evt = arrayFromArguments.apply(null, arguments);

	this._queue.push(evt);

	this.send();

	return false;
};

Bucket.prototype.send = function send () {
	//no pending hooks. How did we get here?
	if (!this._drains.length) return;

	// nothing to send
	if (!this._queue.length) return;

	debug('send', this._queue.length);

	this._drains.shift()(this._queue);
	this._queue = [];
	this._emit('_drained');

	debug('emitted to local', '_drained');

	return this;
};

Bucket.prototype.drain = function drain (callback) {
	var p = proxmis({callback: callback, noError: true});
	var timer, that = this;

	debug('drain');

	var spigot = function (events) {
		// drain fired before the timer, clear the timer.
		if (timer) clearTimeout(timer);
		timer = null;

		debug('draining');

		// pass on the results
		p(events);

		debug('drained');
	};

	timer = setTimeout(function () {
		// Timeout fired before the drain occurred.
		timer = null;

		// Remove the drain listener
		var i = that._drains.indexOf(spigot);
		if (i > -1) that._drains.splice(i, 1);

		debug('drain timeout');
		// Call the proxmis function with an empty array.
		p([]);

		debug('drained empty');
	}, this._options.maxIdle);

	this._drains.push(spigot);
	this.send();

	p.abort = function () {
		if (timer) clearTimeout(timer);

		var i = that._drains.indexOf(spigot);
		if (i > -1) that._drains.splice(i, 1);

		debug('drain aborted');
	};

	return {
		then: p.then,
		catch: p.catch,
		abort: p.abort
	};
};

Bucket.prototype.tap = function tap (emitter) {
	if (!emitter || typeof emitter.emit === 'undefined') throw new TypeError('tap function expected to receive EventEmitter.');
	var pending = this._queue;
	this._queue = [];

	pending.forEach(function (evt) {
		emitter.emit.apply(emitter, evt);
	});

	emitter.emit('_empty');
	debug('emitted to external', '_empty');

	this._emit('_drained');
	debug('emitted to local', '_drained');

	return this;
};


module.exports = Bucket;


/**
* Helper function to convert arguments to an array without triggering de-optimization in V8
* MUST be called via .apply
* @private
* @return {Array<mixed>}
*/
function arrayFromArguments () {
	var len = arguments.length;
	var args = new Array(len);
	for(var i = 0; i < len; ++i) {
		args[i] = arguments[i];
	}
	return args;
}


lib/server.js
var crypto = require('crypto');
var Bucket = require('./bucket.js');
var debug  = require('debug')('long-emitter:server');
var assign = require('lodash.assign');

module.exports = function (options) {
	options = assign({
		querySend: false,
		query: 'streamid',

		idLength: 16,

		expiration: 1000 * 60 * 5, // five minutes
		cleanupPeriod: 1000 * 30   // every 30 seconds
	}, options || {});

	var buckets = {};
	var bucketCount = 0;
	var timeout;

	function makeID () {
		try {
			return crypto.randomBytes(32).toString('hex').substr(0, options.idLength);
		} catch (e) {
			return crypto.pseudoRandomBytes(32).toString('hex').substr(0, options.idLength);
		}
	}

	function create () {
		var id = makeID();
		debug('create', id);

		var emitter = new Bucket(options);
		emitter.id = id;
		emitter.expires = Date.now() + options.expiration;
		emitter.release = function () {
			finish(emitter.id);
		};

		buckets[id] = emitter;
		bucketCount++;

		// click off the expiration checker
		cleanup();
		return emitter;
	}

	function get (id) {
		debug('get', id);
		if (buckets[id]) {
			//reset expiration time
			buckets[id].expires = Date.now() + options.expiration;
		}
		return buckets[id];
	}

	function close (id) {
		if (typeof id === 'object') id = id.id;

		debug('close', id);
		var bucket = buckets[id];
		if (!bucket) return server;

		bucket.removeAllListeners();
		delete buckets[id];

		bucketCount--;
		if (!bucketCount && timeout) {
			// nothing left to clean up
			clearTimeout(timeout);
			timeout = null;
		}

		return server;
	}

	function finish (id) {
		if (typeof id === 'object') id = id.id;

		debug('finish', id);
		var bucket = buckets[id];
		if (!bucket) return server;
		
		bucket.once('_drained', function () {
			close(id);
			debug('finished', id);
		});
	}

	function cleanup () {
		// check if there are existing buckets and no cleanup already pending
		if (!bucketCount || timeout) return;

		debug('cleanup');
		timeout = setTimeout(function () {
			timeout = null;

			var ids = Object.keys(buckets);
			var now = Date.now();

			// loop through all buckets and remove those which have expired
			ids.forEach(function (id) {
				if (buckets[id].expires < now) {
					close(id);
				}
			});

			debug('cleanedup');
			cleanup();
		}, options.cleanupPeriod);

		return server;
	}

	function clear () {
		debug('clear');
		if (timeout) {
			clearTimeout(timeout);
			timeout = null;
		}

		buckets = {};
		bucketCount = 0;

		return server;
	}

	function param (req, res, next, id) {
		debug('param', id);
		var emitter = get(id);

		if (!emitter) return next();

		var hook = emitter.drain(function (events) {
			debug('param drained', id);
			req.removeListener('close', hook.abort);
			res.locals.events = events;
			next();
		});

		// if the request disconnects before completion, abort the drain.
		req.on('close', hook.abort);
	}

	function query (req, res, next) {
		var id;
		if (typeof options.query === 'string') {
			id = req.query[options.query];
		} else if (typeof options.query === 'function') {
			id = options.query(req);
		}

		debug('query', id);

		var emitter = get(id);

		if (!emitter) {
			if (options.querySend) {
				return next('route');
			} else {
				return next();
			}
		}
		
		var hook = emitter.drain(function (events) {
			debug('query drained', id);
			req.removeListener('close', hook.abort);
			if (options.querySend) {
				res.json(events);
			} else {
				res.locals.events = events;
				next();
			}
		});

		// if the request disconnects before completion, abort the drain.
		req.on('close', hook.abort);

	}

	var server = {
		create: create,
		get: get,
		close: close,
		finish: finish,
		param: param,
		query: query,
		clear: clear
	};

	return server;
};

module.exports.Bucket = Bucket;

lib/manager.js
var crypto = require('crypto');
var Bucket = require('./bucket.js');
var debug  = require('debug')('long-emitter:manager');
var assign = require('lodash.assign');

module.exports = function (options) {
	options = assign({
		querySend: false,
		query: 'streamid',

		idLength: 16,

		expiration: 1000 * 60 * 5, // five minutes
		cleanupPeriod: 1000 * 30   // every 30 seconds
	}, options || {});

	var buckets = {};
	var bucketCount = 0;
	var timeout;

	function makeID () {
		try {
			return crypto.randomBytes(32).toString('hex').substr(0, options.idLength);
		} catch (e) {
			return crypto.pseudoRandomBytes(32).toString('hex').substr(0, options.idLength);
		}
	}

	function create () {
		var id = makeID();
		debug('create', id);

		var emitter = new Bucket(options);
		emitter.id = id;
		emitter.expires = Date.now() + options.expiration;
		emitter.release = function () {
			finish(emitter.id);
		};

		buckets[id] = emitter;
		bucketCount++;

		// click off the expiration checker
		cleanup();
		return emitter;
	}

	function get (id) {
		debug('get', id);
		if (buckets[id]) {
			//reset expiration time
			buckets[id].expires = Date.now() + options.expiration;
		}
		return buckets[id];
	}

	function close (id) {
		if (typeof id === 'object') id = id.id;

		debug('close', id);
		var bucket = buckets[id];
		if (!bucket) return manager;

		bucket.removeAllListeners();
		delete buckets[id];

		bucketCount--;
		if (!bucketCount && timeout) {
			// nothing left to clean up
			clearTimeout(timeout);
			timeout = null;
		}

		return manager;
	}

	function finish (id) {
		if (typeof id === 'object') id = id.id;

		debug('finish', id);
		var bucket = buckets[id];
		if (!bucket) return manager;
		
		bucket.once('_drained', function () {
			close(id);
			debug('finished', id);
		});
	}

	function cleanup () {
		// check if there are existing buckets and no cleanup already pending
		if (!bucketCount || timeout) return;

		debug('cleanup');
		timeout = setTimeout(function () {
			timeout = null;

			var ids = Object.keys(buckets);
			var now = Date.now();

			// loop through all buckets and remove those which have expired
			ids.forEach(function (id) {
				if (buckets[id].expires < now) {
					close(id);
				}
			});

			debug('cleanedup');
			cleanup();
		}, options.cleanupPeriod);

		return manager;
	}

	function clear () {
		debug('clear');
		if (timeout) {
			clearTimeout(timeout);
			timeout = null;
		}

		buckets = {};
		bucketCount = 0;

		return manager;
	}

	function param (req, res, next, id) {
		debug('param', id);
		var emitter = get(id);

		if (!emitter) return next();

		var hook = emitter.drain(function (events) {
			debug('param drained', id);
			req.removeListener('close', hook.abort);
			res.locals.events = events;
			next();
		});

		// if the request disconnects before completion, abort the drain.
		req.on('close', hook.abort);
	}

	function query (req, res, next) {
		var id;
		if (typeof options.query === 'string') {
			id = req.query[options.query];
		} else if (typeof options.query === 'function') {
			id = options.query(req);
		}

		debug('query', id);

		var emitter = get(id);

		if (!emitter) {
			if (options.querySend) {
				return next('route');
			} else {
				return next();
			}
		}
		
		var hook = emitter.drain(function (events) {
			debug('query drained', id);
			req.removeListener('close', hook.abort);
			if (options.querySend) {
				res.json(events);
			} else {
				res.locals.events = events;
				next();
			}
		});

		// if the request disconnects before completion, abort the drain.
		req.on('close', hook.abort);

	}

	var manager = {
		create: create,
		get: get,
		close: close,
		finish: finish,
		param: param,
		query: query,
		clear: clear
	};

	return manager;
};

module.exports.Bucket = Bucket;